scripts/zkai_native_attention_mlp_adapter_opening_geometry_budget_gate.py

```
@@ -0,0 +1,895 @@
+    if json_path is not None:
+        json_path = require_output_path(json_path, ".json")
+        write_text_atomically(json_path, json.dumps(payload, indent=2, sort_keys=True) + "\n")
+    if tsv_path is not None:
+        tsv_path = require_output_path(tsv_path, ".tsv")
+        write_text_atomically(tsv_path, build_tsv_text(payload))
+
+
+def require_output_path(path: pathlib.Path, suffix: str) -> pathlib.Path:
+    candidate = path if path.is_absolute() else ROOT / path
+    if candidate.suffix != suffix:
+        raise AdapterOpeningGeometryBudgetError(f"output path must use {suffix} suffix: {path}")
```

Ruff (0.15.12)

[warning] Avoid specifying long messages outside the exception class

(TRY003)